Makefile

```diff
@@ -1,85 +1,92 @@
+#########################################
+build: init \
+       $(BINDIR)/controller $(BINDIR)/registry $(BINDIR)/k8s-broker \
+       $(BINDIR)/service-catalog $(BINDIR)/user-broker
+
+$(BINDIR)/controller: pkg/controller/catalog
+	$(DOCKER) $(GO_BUILD) -o $@ $(SC_PKG)/$^
+
+$(BINDIR)/registry: contrib/registry
+	$(DOCKER) $(GO_BUILD) -o $@ $(SC_PKG)/$^
 
-# Use this target when you want to build everything using docker containers.
-# Good for cases when you don't have the tools installed (like glide or go).
-docker-all: .scBuildImage
-	docker run --rm -ti \
-	  -v $(PWD):/go/src/github.com/kubernetes-incubator/service-catalog \
-	  scbuildimage \
-	  make .dockerInit all
-
-# .dockerInit tells us if our vendor stuff if out of date or not.
-# And if so we'll run init under our docker build.  For non-docker builds
-# it is assumed you'll run "make init" manually.
-.dockerInit: glide.yaml
-	make init
-	echo > .dockerInit
-
-# .scBuildImage is used to know when the docker image ("scbuildimage") is out
-# of date with the Dockerfile.
+$(BINDIR)/k8s-broker: contrib/broker/k8s
+	$(DOCKER) $(GO_BUILD) -o $@ $(SC_PKG)/$^
+
+$(BINDIR)/service-catalog: cmd/service-catalog
+	$(DOCKER) $(GO_BUILD) -o $@ $(SC_PKG)/$^
+
+$(BINDIR)/user-broker: contrib/broker/k8s
+	$(DOCKER) $(GO_BUILD) -o $@ $(SC_PKG)/$^
+
+# Some prereq stuff
+###################
+init: .scBuildImage .init
+
+# .init is used to know when some Glide dependencies are out of date
+.init: glide.yaml
+	$(DOCKER) glide --home .glide install --strip-vendor
+	echo > $@
+
+# .scBuildImage tells when the docker image ("scbuildimage") is out of date
 .scBuildImage: hack/Dockerfile
 	docker build -t scbuildimage - < hack/Dockerfile
-	echo > .scBuildImage
+	echo > $@
 
-# Build the same target recursively in all directories.
-$(SUB): %.sub:
-	$(ECHO) for dir in $(DIRS); do $(MAKE) --no-print-directory -C "$${dir}" $* || exit $$? ; done
+# Util targets 
+##############
 
-init:
-	$(ECHO) glide install --strip-vendor
+verify: init
+	@echo Running verify:
+	@for i in contrib pkg; do \
+      $(DOCKER) gofmt -l $$i ; \
+	done > .out 2>&1
+	@bash -c '[ "`cat .out`" == "" ] || \
+	  (echo -e "\n*** Please 'gofmt' the following:" ; cat .out ; echo ; false)'
+	@rm .out
+	@for i in $(SRC_PKGS); do \
+	  $(DOCKER) golint --set_exit_status $$i/... && \
+	  $(DOCKER) go vet $$i ; \
+	done || false
 
-format:
-	$(ECHO) gofmt -w -s $(addprefix ./,$(DIRS))
+format: init
+	$(DOCKER) gofmt -w -s contrib pkg
 
-coverage:
-	$(ECHO) $(ROOT)/hack/coverage.sh --html "$(COVERAGE)" $(addprefix ./,$(DIRS))
+coverage: init
+	$(DOCKER) hack/coverage.sh --html "$(COVERAGE)" $(addprefix ./,$(TEST_DIRS))
 
-.PHONY: apiserver
-apiserver:
-	go install -v github.com/kubernetes-incubator/service-catalog/cmd/service-catalog
-	go build -v -o $(BINDIR)/apiserver cmd/service-catalog/server.go
+test: 
+	@echo Running tests:
+	@for i in $(SRC_PKGS); do \
+	  $(DOCKER) go test $$i ; \
+	done
+
+clean:
+	rm -rf $(BINDIR)
```

Comment:
what principle do you apply when deciding whether to use `@` or not? (compare lines 70 and 89, and 84 for example)